### PR TITLE
RCTProfile: Use C atomics instead of OSAtomic

### PR DESCRIPTION
**Pull request by Adlai-Holler**
This will save us precious microseconds and it's prettier to look at.

Test plan: Try the profiler, rely on CI.

### CODE REVIEW

**Comment by shergin** [comment]
cc @javache

**Comment by javache** [comment]
Can you share some numbers from profiling this? Or is this based on intuition?

File: React/Profiler/RCTProfile.m [hunk 1]
```diff
@@ -42,8 +42,7 @@
 
 #pragma mark - Variables
 
-// This is actually a BOOL - but has to be compatible with OSAtomic
-static volatile uint32_t RCTProfileProfiling;
+static atomic_bool RCTProfileProfiling = ATOMIC_VAR_INIT(NO);
```
Comment (javache) [inline React/Profiler/RCTProfile.m]: Can you mark this as `constexpr` so we can verify at compile-time this doesn't add a global initalizer?

Comment (Adlai-Holler) [inline React/Profiler/RCTProfile.m]: Hmm `constexpr` isn't available except in C++. We can make this `RCTProfile.mm` but that's a bigger change and we'll lose a tiny bit of efficiency (due to ARC exception safety). What do?

**Comment by Adlai-Holler** [comment]
@javache It's just intuition since `OSAtomic*` are function calls and `atomic_*` are inlined down to one or two instructions. I doubt the performance difference will be measurable for our current uses. But the main benefit here IMO is that `OSAtomic` is deprecated ( https://github.com/facebook/react-native/pull/15278#issuecomment-318938721 ) plus I just love language features.

File: React/Profiler/RCTProfile.m (hunk 1 above)
Comment (javache) [inline React/Profiler/RCTProfile.m]: Our internal setup is to have `-fno-exceptions` for most objective-c code and opt-in to the codepaths that need it. Maybe we can do that here too?

Comment (Adlai-Holler) [inline React/Profiler/RCTProfile.m]: Hmm turns out we can't use `constexpr` because it implies `const` and we need to mutate this value.

**Comment by Adlai-Holler** [comment]
I don't think `constexpr` is doable, but I've rebased the diff for good measure

File: React/Profiler/RCTProfile.m (hunk 1 above)
Comment (javache) [inline React/Profiler/RCTProfile.m]: Hmm ok. I think we compile by default with -Wglobal-constructors, so that should catch it too.

**Pull request closed without merging**